Add support for SEGGER RTT stdio

immesys opened the pull request:
A platform we are developing for the SAMR21 does not possess a UART for STDIO. Instead we use SEGGER's Real Time Transfer or RTT. It works by maintaining ringbuffers in RAM that are read and written to over JTAG. It might be useful in general because it is much faster than a UART.

Some things to note: the module is designed to replace uart_stdio (and has the same signatures) so that newlib will use it without any modification if it is included in USEMODULE. By default, stdin is disabled, because there is no interrupt when input data appears - memory just changes. Therefore you have to poll the input ringbuffer. If the platform is power constrained, and is trying to minimize its duty cycle, you certainly don't want to wake up and poll. If you are connected to a debugger, however, then it probably doesn't matter if you poll so your application can call `rtt_stdio_enable_stdin()` . I considered inspecting the DHCSR register to determine the presence of a debugger automatically but only certain Cortex-M processors have that feature, and the rtt-stdio should in theory work with any microcontroller supported by SEGGER.

The code itself is only slightly modified from SEGGER's reference RTT implementation. I cleaned it up to match the RIOT style guides (I think) and also remove some operating modes that are not needed for plain stdio.

I have tested this on both our platform and on a SAMR21 XPRO. You will need to connect to the jtag header directly, I don't think SEGGERs RTT software will work with the Atmel EDBG chip.

If you want to test it, there is a demo app here: https://github.com/immesys/RiSyn/tree/samr21xpro-rtt-demo/examples/samr21xpro-rtt-demo
You will need to run JLinkGDBServer and then run JLinkRTTClient (they might have different names on other operating systems, I have only tested on Linux)

This is my first nontrivial pull request, and I intend to do a bunch more, so I would appreciate any constructive feedback with regards to style, contribution policy etc.

immesys commented:
Just a thought, I probably need to specify xtimer as a dependency for rtt_stdio. I'll add that in soon

DipSwitch commented:
Nice one! :)

immesys commented:
After using this a bit for more complicated applications, two things have come up:

Firstly, I forgot to modify the application to enable stdin. I think other people might do this too. Which is more common, that people will use stdin or that they won't and also care about the wasted cpu time / power? If using stdin is more common, I think enabling it by default is a better idea. It can be disabled via a #define, unless that is against RIOT style guides?

Secondly, the shell has a habit of printing out large volumes of text all at once (like the help command). In the case where the developer knows there is an RTT client attached, it would be nice to block in uart_stdio_write (like a uart does) so that we do not drop chunks of text. The alternative is increasing the TX buffer. I think having the option for blocking stdout might be useful to some people, so I will add it as an option that can be turned on with a #define. I don't think it is wise to have it on by default because then every thread with a printf in it will hang forever waiting for an RTT client to read the ringbuffer.

immesys commented:
Rebased and squashed. This is not quite ready for merging yet, there is a corner case I am busy investigating.

DipSwitch commented:
What is the corner case? I would like to play with SEGGER SystemView / tracing in RIOT. So maybe I could help?

immesys commented:
The blocking stdout was not working for me when I briefly tried it recently. When I initially made the PR I could swear it worked. I haven't had time to look at it though. It should be in a place where you can test it. Please let me know how it works for you!

thomaseichinger commented on sys/rtt_stdio/rtt_stdio.c:
```diff
@@ -0,0 +1,282 @@
+ * @brief SEGGER's ring buffer implementation
+ */
+typedef struct {
+    const char* channel_name; // Optional name. Standard names so far are:
```
Here and all following, please use `/* C-style comments */`.

thomaseichinger commented:
@immesys Some minor comments, code seems fine I couldn't test yet though.

kYc0o commented:
Feature freeze -> postponed.

OlegHahm commented:
Why did you close?

immesys commented:
Interesting. I forgot that I didn't follow this PR through. We are using this a lot and the code works fine, I just rebased into a new branch and deleted the old one and I guess it closed this. I don't think I addressed @thomaseichinger's suggestions, so I will do that and open a new PR with the rebased branch.
